katdal/categorical.py

```diff
@@ -580,8 +580,71 @@ def concatenate_categorical(split_data, **kwargs):
-                          greedy_values=None, initial_value=None, transform=None, allow_repeats=False, **kwargs):
+def _single_event_per_dump(events, greedy):
+    """Ensure that each dump is associated with a single sensor event.
+
+    This generates a sequence of cleaned-up sensor events (represented by
+    indices into the original `events` sequence), which ensures that each dump
+    is associated with a single event. When there are multiple events inside
+    a dump, pick the final one. In addition, some sensor values designated as
+    "greedy" will override non-greedy ones and grab a dump even if it is not
+    the final value. In this scenario, move the final (non-greedy) event to the
+    next dump by modifying its dump index in the `events` parameter. The
+    generator returns up to *N* events but not the special terminal event.
+
+    Parameters
+    ----------
+    events : mutable sequence of non-negative ints, length *N* + 1
+        Monotonic sequence of dump indices associated with each sensor event.
+        The last event is one past the last dump (i.e. the total number of
+        dumps). Be aware that this parameter is mutated by the function.
+    greedy : sequence of bool, length *N*
+        Flags indicating whether the sensor value at a given event is "greedy"
+
+    Yields
+    ------
+    event : non-negative int
+        Index into `events` sequence of the next cleaned up event (does not
+        yield the one-past-last-dump terminal event)
+
+    """
+    # This will generate consecutive event indices with associated dump indices
+    dumps = enumerate(events)
+    # The previous winning event is the dominant event in the previous dump
+    previous_dump, previous_winning_event = 0, 0
+    while True:
+        current_event, current_dump = next(dumps)
+        # At the start of a new dump, process the events of the previous dump
+        if current_dump > previous_dump:
+            # This previous event segment is assumed to straddle dump boundary
+            # ASSERT: current_event >= 1 since we are now at least at 2nd dump
+            event_at_dump_start = current_event - 1
+            # The normal victory condition is to be the final event in the dump
+            if not greedy[previous_winning_event]:
+                previous_winning_event = event_at_dump_start
+            winning_dump = events[previous_winning_event]
+            # Only yield winning event in immediate past to avoid duplicates
+            if previous_dump <= winning_dump and winning_dump < current_dump:
+                yield previous_winning_event
+            # If winning event was greedy and final event was non-greedy,
+            # push final event to the start of next dump and yield if it is
+            # the only event in that dump (otherwise it has to fight it out...)
+            if event_at_dump_start != previous_winning_event:
+                # NB: This modifies `events`! It simplifies bookkeeping.
```

Comment:
Yes. The "event at dump start" will either be resolved immediately or affect at most the next iteration of `dumps`. There it will be handled specially, i.e. it will explicitly index `events` to retrieve its now modified dump index. I'm effectively modifying the past of the `dumps` iterator, not the present or future.